### ISSUE TITLE
Beta: sign-in links do not expire after 15 minutes and are reusable

### ISSUE THREAD

**Issue by leonfeng** [freeCodeCamp/freeCodeCamp#16912]
#### Issue Description

I am able to sign in using the links well after 15 minutes (and in one case, 2 hours). And I can do so repeatedly using the same link in different browsers (private mode) and on different devices. Reproducible on my local copy of freeCodeCamp and on beta.freecodecamp.org.

#### Browser Information

* Browser Name, Version: Chrome, Firefox, Safari
* Operating System: macOS
* Mobile, Desktop, or Tablet: mobile and desktop

**Comment by leonfeng** [comment]
Wait… just checked my MongoDB records. The auth tokens' TTLs are all set to 900000. Is that intentional?

**Comment by ojongerius** [comment]
Interesting, 900000 ms would be 15 minutes. Are they not miliseconds? If they are seconds those links would stay valid for 15000 minutes instead, that's more than 10 days 😲

**Comment by ojongerius** [comment]
Ok, I can confirm that:

* authentication links are not deleted after use
* authentication links are valid for longer than the ttl

My guess is that they will never expire, as I do not see the code for it, but I'm pretty new to the code base.

**Comment by raisedadead** [comment]
Hi @leonfeng 

Nice find.

I guess, the last discussions we had about an year ago that MongoDB should be doing this for us. One of the causes this could be the TTL expiration thread needs to be configured explicitly?

But, I guess we would have to have this in to a message queue, and schedule it.

I will investigate further.

**Comment by raisedadead** [comment]
Or, maybe we simply missed the invalidation in a recent refactoring.

**Comment by ojongerius** [comment]
I'll just drop here what I cam across trying to unravel this little puzzle, hope it's helpful:

Intention to validate:
https://github.com/freeCodeCamp/freeCodeCamp/blob/bce8e0a5e93dca57d7734de5b0d6abedd0517272/server/boot/authentication.js#L140

Intention to destroy:
https://github.com/freeCodeCamp/freeCodeCamp/blob/bce8e0a5e93dca57d7734de5b0d6abedd0517272/server/boot/authentication.js#L155

I noticed that validation of length and existence of the token does happen, I'm not familiar with the code base and frameworks but that seems to happen in `passwordlessGetValidators`:

https://github.com/freeCodeCamp/freeCodeCamp/blob/bce8e0a5e93dca57d7734de5b0d6abedd0517272/server/boot/authentication.js#L69

I guess a shortcut is to add TTL validation in `passwordlessGetValidators`, but as you still need to solve the removal of used tokens, I'd suggest fixing it properly.

**Comment by scissorsneedfoodtoo** [comment]
Hi all. I saw that @raisedadead submitted a PR to migrate authentication to Auth0, but I thought I'd comment here.

Did some searching through the Mongo docs for getting TTL to work. Seems like Mongo will automatically remove documents, but an index needs to be created on the collection first. To get the email authentication links to expire correctly, I had to run mongo, use freecodecamp, then run `db.AuthToken.createIndex( { "created": 1 }, { expireAfterSeconds: 900 } )`. After that, the links would be removed and would no longer work after 15 minutes, give or take a minute -- the background task that removes documents runs every 60 seconds, apparently.

One thing I noticed is that the authentication token is not always destroyed when the user clicks on the link and logs in. I'm not sure what is causing the issue, but it seems to be destroyed about half the time. So a user could click on an email link, log in, sign out, then log in multiple times until the token is destroyed or ~15 minutes passes.

Another potential issue I saw is that the db.AccessToken collection seems to grow every time a user logs in with a new authentication token. Not sure if that's supposed to happen, but I thought it worth mentioning. Also, the `ttl` field in the documents in db.AccessToken reads `NaN`. Again, not sure, but maybe it's supposed to be set by this line?:

https://github.com/freeCodeCamp/freeCodeCamp/blob/a2076ce1ec96e5c333964f20dc20cc9621501f37/common/models/user.js#L181
